source/fab/tasks/fortran.py

```diff
@@ -464,80 +486,110 @@ def run(self):
+                 flags: List[str],
+                 workspace: Path):
+        self._preprocessor = preprocessor
+        self._flags = flags
+        self._workspace = workspace
+
+    def run(self, artifacts: List[Artifact]) -> List[Artifact]:
 
-class FortranPreProcessor(SingleFileCommand):
+        if len(artifacts) == 1:
+            artifact = artifacts[0]
+        else:
+            msg = ('Fortran Preprocessor expects only one Artifact, '
+                   f'but was given {len(artifacts)}')
+            raise TaskException(msg)
 
-    @property
-    def as_list(self) -> List[str]:
-        base_command = ['cpp', '-traditional-cpp', '-P']
-        file_args = [str(self._filename), str(self.output[0])]
-        return base_command + self._flags + file_args
+        artifact = artifacts[0]
+        command = [self._preprocessor]
+        command.extend(self._flags)
+        command.append(str(artifact.location))
 
-    @property
-    def output(self) -> List[Path]:
-        return [self._workspace /
-                self._filename.with_suffix('.f90').name]
+        output_file = (self._workspace /
+                       artifact.location.with_suffix('.f90').name)
+        command.append(str(output_file))
 
+        subprocess.run(command, check=True)
 
-class FortranCompiler(Command):
+        return [Artifact(output_file,
+                         artifact.filetype,
+                         Raw)]
+
+
+class FortranCompiler(Task):
 
     def __init__(self,
-                 filename: Path,
-                 workspace: Path,
+                 compiler: str,
                  flags: List[str],
-                 prerequisites: List[Path]):
-        super().__init__(workspace, flags)
-        self._filename = filename
-        self._prerequisites = prerequisites
-
-    @property
-    def as_list(self) -> List[str]:
-        base_command = ['gfortran',
-                        '-c',
-                        '-J' + str(self._workspace),
-                        ]
-        file_args = [str(self._filename),
-                     '-o',
-                     str(self.output[0]),
-                     ]
-        return base_command + self._flags + file_args
-
-    @property
-    def input(self) -> List[Path]:
-        return self._prerequisites + [self._filename]
-
-    @property
-    def output(self) -> List[Path]:
-        object_file = (
-            self._workspace / self._filename.with_suffix('.o').name)
-        return [object_file]
-
-
-class FortranLinker(Command):
+                 workspace: Path):
+        self._compiler = compiler
+        self._flags = flags
+        self._workspace = workspace
+
+    def run(self, artifacts: List[Artifact]) -> List[Artifact]:
+
+        if len(artifacts) == 1:
+            artifact = artifacts[0]
+        else:
+            msg = ('Fortran Compiler expects only one Artifact, '
+                   f'but was given {len(artifacts)}')
+            raise TaskException(msg)
```

Comment:
Given that all these tasks are doing this check there may be space to hoist the functionality into a an abstract parent.

Reply:
Can I make the `@abstracmethod` in the `Task` parent class for `run` have this logic and then get each class to call `artifact = super().run(artifacts)` or would that be mis-using the way abstract methods are meant to be used?  Either that or I could add a class-level attribute which stores something like `input_type = 'Single'` (or `'Multiple'`) and provide a method the classes call like `artifact = self.check_input(artifacts)`... though that feels a little but cumbersome?  (Did you have a specific way in mind?)

Reply:
Calling your parent implementation is absolutely a normal part of OO coding. The child implementation is modifying the behaviour of its parent in some way.